core/regions/src/main/java/software/amazon/awssdk/regions/servicemetadata/EnhancedS3ServiceMetadata.java

```diff
@@ -0,0 +1,100 @@
+ */
+
+package software.amazon.awssdk.regions.servicemetadata;
+
+import java.net.URI;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.core.SdkSystemSetting;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.profiles.ProfileFileSystemSetting;
+import software.amazon.awssdk.profiles.ProfileProperty;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.regions.ServiceMetadata;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * Decorator metadata class for S3 to allow customers to opt in to using the
+ * regional S3 us-east-1 endpoint instead of the legacy
+ * {@code s3.amazonaws.com} when specifying the us-east-1 region.
+ */
+@SdkPublicApi
+public final class EnhancedS3ServiceMetadata implements ServiceMetadata {
+    private static final Logger log = Logger.loggerFor(EnhancedS3ServiceMetadata.class);
+    private static final String REGIONAL_SETTING = "regional";
+    private static final S3ServiceMetadata S3_SERVICE_METADATA = new S3ServiceMetadata();
+
+    private volatile boolean profileSettingChecked = false;
+    private volatile String profileSetting = null;
+
+    @Override
+    public URI endpointFor(Region region) {
+        if (Region.US_EAST_1.equals(region) && !useUsEast1RegionalEndpoint()) {
+            return URI.create("s3.amazonaws.com");
+        }
+        return S3_SERVICE_METADATA.endpointFor(region);
+    }
+
+    @Override
+    public Region signingRegion(Region region) {
+        return S3_SERVICE_METADATA.signingRegion(region);
+    }
+
+    @Override
+    public List<Region> regions() {
+        return S3_SERVICE_METADATA.regions();
+    }
+
+    private boolean useUsEast1RegionalEndpoint() {
+        String env = envVarSetting();
+
+        if (env != null) {
+            return REGIONAL_SETTING.equalsIgnoreCase(env);
+        }
+
+        String profile = profileFileSetting();
+
+        if (profile != null) {
+            return REGIONAL_SETTING.equalsIgnoreCase(profile);
+        }
+
+        return false;
+    }
+
+    private static String envVarSetting() {
+        return SdkSystemSetting.AWS_S3_US_EAST_1_REGIONAL_ENDPOINT.getStringValue().orElse(null);
+    }
+
+    private String profileFileSetting() {
+        if (!profileSettingChecked) {
+            synchronized (this) {
+                if (!profileSettingChecked) {
+                    try {
+                        String profileName = ProfileFileSystemSetting.AWS_PROFILE.getStringValueOrThrow();
+                        ProfileFile profileFile = ProfileFile.defaultProfileFile();
+                        profileSetting = profileFile.profile(profileName)
+                                .flatMap(p -> p.property(ProfileProperty.S3_US_EAST_1_REGIONAL_ENDPOINT))
+                                .orElse(null);
+                    } catch (Throwable t) {
+                        log.warn(() -> "Unable to load config file", t);
```

Comment:
Should we use `debug` instead? Would this be logged for every request?

Reply:
No, it's only checked once https://github.com/aws/aws-sdk-java-v2/blob/442ec96fe12b5398a034bd693760c4fe853da8e6/core/regions/src/main/java/software/amazon/awssdk/regions/servicemetadata/EnhancedS3ServiceMetadata.java#L95

Reply:
It wouldn't log on each request AFAIK since the generated `GeneratedServiceMetadataProvider` class that instantiates this is a is a singleton instance.

Reply:
My concern is that customers not using the profiles will get concerned when they see the warn message. I recall we've had customers reporting similar issue with other profile config before.

Reply:
There shouldn't be an exception if they don't have any profile files:

https://github.com/aws/aws-sdk-java-v2/blob/4e4bebd8e17ba4651b5d0ee6d667a80aa5fc2601/core/profiles/src/main/java/software/amazon/awssdk/profiles/ProfileFile.java#L134-L148

The created `ProfileFile` will essentially be empty.

Reply:
Gotcha, then we're good